text/0000-introduce empty ranges.md

```diff
@@ -0,0 +1,425 @@
+
+> ```pony
+> class Range[A: (Real[A] val & Number) = USize] is Iterator[A]
+>   ..
+> let _min: A
+>   let _max: A
+>   let _inc: A
+>   let _forward: Bool
+>   let _infinite: Bool
+>   var _idx: A
+> 
+>   new create(min: A, max: A, inc: A = 1) =>
+>     _min = min
+>     _max = max
+>     _inc = inc
+>     _idx = min
+>     _forward = (_min < _max) and (_inc > 0)
+>     let is_float_infinite =
+>       iftype A <: FloatingPoint[A] then
+>         _min.nan() or _min.infinite()
+>           or _max.nan() or _max.infinite()
+>           or _inc.nan() or _inc.infinite()
+>       else
+>         false
+>       end
+>     _infinite =
+>       is_float_infinite
+>         or ((_inc == 0) and (min != max))    // no progress
+>         or ((_min < _max) and (_inc < 0)) // progress into other directions
+>         or ((_min > _max) and (_inc > 0)) 
+> ```
+
+This RFC partitions the set of currently *infinite* `Range` cases into those now considered *empty*, and a very small set of those that remain *infinite*.
+
+For this, two criteria are postulated that are required for a Range to not be empty. Numeric values including the floating point values `+Inf` and `-Inf` can be meaningfully compared in expressions like `min > max` or `max > min`. Based on this, an obvious absence of *progress* of the iteration from min towards max based on the sign of the Range parameters can be tested by `((min <= max) and (step <= 0)) or ((min >= _max) and (step >= 0))` ("no-progress" criterion) for finite and infinite bounds as well as finite and infinite values of step (note that if either of the `min` or `max` parameters is `NaN` such expressions are always `false` and this special case is discussed further below). Here, besides the obvious sign-incompatibility of the three Range parameters, the cases of equal bounds `min == max` and `step == 0` are included, since the former wouldn't satisfy `[min, min)` (and also doesn't require any progress) and the latter prevents any progress. In the current implementation of Range, a somewhat similar couple of expressions are used with the comment `// no progress` and `// progress into other directions`, but they are treated as a sufficient condition for *infinite*. Here, this criterion is used as part of a test for empty Ranges because it allows us to know in which cases the generation of values between min and max is impossible, irrespective of the magnitude of the bounds or steps, in other words, no valid *trajectory* from min to max does exists with the provided parameters. The second important criterion used here is that for a Range iterators to be not empty, it must only produce finite values even when infinite bounds are specified, and if that is not possible, the Range is again *empty*.
```

Comment:
Yes, I should try to simplify it a lot more. I realize it is too wordy and in places too hard to follow.

Reply:
Thank you. I think there is a general tendency by most folks when describing mathematical concepts in plain language to try to keep caveats in the reader's mind which eventually leads to wordiness. Simplifying and reducing this wordiness can significantly help readability -- even if we sometimes have to remind ourselves of the caveats that otherwise would have been kept in place.

Reply:
I rewrote quite a bit of this in the hope to make the discussion more concise.